File: Week6 (Graph [SSSP, APSP, BFS, DFS])/All Pair Shortest Path (APSP)/cmy-apsp-johnson.py
import heapq
import sys

def bellman_ford(source, graph, n):
    h = [float('inf')] * n
    h[source] = 0
    
    for _ in range(n - 1):
        for u in range(n):
            for v, weight in graph[u]:
                if h[u] != float('inf') and h[u] + weight < h[v]:
                    h[v] = h[u] + weight
    
    return h

def dijkstra(source, graph, n):
    dist = [float('inf')] * n
    dist[source] = 0
    
    pq = [(0, source)]
    while pq:
        current_dist, u = heapq.heappop(pq)
        
        if current_dist > dist[u]:
            continue
        
        for v, weight in graph[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                heapq.heappush(pq, (dist[v], v))
    
    return dist

def johnson(graph, n):
    modified_graph = [edges[:] for edges in graph]
    h = bellman_ford(n, modified_graph + [[]], n)
    
    for u in range(n):
        for i in range(len(modified_graph[u])):
            v, weight = modified_graph[u][i]
            modified_graph[u][i] = (v, weight + h[u] - h[v])
    
    all_distances = []
    for u in range(n):
        dist = dijkstra(u, modified_graph, n)
        all_distances.append([dist[v] + h[v] - h[u] if dist[v] < float('inf') else float('inf') for v in range(n)])
    
    return all_distances

def main():
    input = sys.stdin.read
    data = list(map(int, input().split()))
    
    n = data[0]
    m = data[1]
    
    graph = [[] for _ in range(n)]
    index = 2
    for _ in range(m):
        u = data[index] - 1
        v = data[index + 1] - 1
        w = data[index + 2]
        graph[u].append((v, w))
        index += 3
    
    all_distances = johnson(graph, n)
    
    for row in all_distances:
        print(" ".join("INF" if dist == float('inf') else str(dist) for dist in row))

if __name__ == "__main__":
    main()

File: Week6 (Graph [SSSP, APSP, BFS, DFS])/All Pair Shortest Path (APSP)/cmy-apsp-johnson.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>

using namespace std;

#define INF INT_MAX

void bellmanFord(int source, const vector<vector<pair<int, int>>>& graph, vector<int>& h) {
    int n = graph.size();
    fill(h.begin(), h.end(), INF);
    h[source] = 0;
    
    for (int i = 0; i < n - 1; ++i) {
        for (int u = 0; u < n; ++u) {
            for (const auto& edge : graph[u]) {
                int v = edge.first;
                int weight = edge.second;
                
                if (h[u] != INF && h[u] + weight < h[v]) {
                    h[v] = h[u] + weight;
                }
            }
        }
    }
}

void dijkstra(int source, const vector<vector<pair<int, int>>>& graph, vector<int>& dist) {
    int n = graph.size();
    fill(dist.begin(), dist.end(), INF);
    dist[source] = 0;
    
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, source});
    
    while (!pq.empty()) {
        int u = pq.top().second;
        int dist_u = pq.top().first;
        pq.pop();
        
        if (dist_u > dist[u]) continue;
        
        for (const auto& edge : graph[u]) {
            int v = edge.first;
            int weight = edge.second;
            
            if (dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
                pq.push({dist[v], v});
            }
        }
    }
}

void johnson(const vector<vector<pair<int, int>>>& graph, vector<vector<int>>& all_distances) {
    int n = graph.size();
    vector<vector<pair<int, int>>> modified_graph = graph;
    vector<int> h(n, INF);
    
    vector<vector<pair<int, int>>> temp_graph(n + 1);
    for (int u = 0; u < n; ++u) {
        temp_graph[n].push_back({u, 0});
    }
    
    temp_graph.push_back(vector<pair<int, int>>());
    for (int u = 0; u < n; ++u) {
        for (const auto& edge : graph[u]) {
            temp_graph[u].push_back(edge);
        }
    }
    
    bellmanFord(n, temp_graph, h);
    
    for (int u = 0; u < n; ++u) {
        modified_graph[u].clear();
        for (const auto& edge : graph[u]) {
            int v = edge.first;
            int weight = edge.second;
            modified_graph[u].push_back({v, weight + h[u] - h[v]});
        }
    }
    
    all_distances.resize(n);
    for (int i = 0; i < n; ++i) {
        all_distances[i].resize(n);
        dijkstra(i, modified_graph, all_distances[i]);
        for (int j = 0; j < n; ++j) {
            if (all_distances[i][j] != INF) {
                all_distances[i][j] += h[j] - h[i];
            }
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m; // Number of nodes and edges
    
    vector<vector<pair<int, int>>> graph(n);
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        cin >> u >> v >> w; // Edge from u to v with weight w
        graph[u - 1].push_back({v - 1, w});
    }
    
    vector<vector<int>> all_distances;
    johnson(graph, all_distances);
    
    for (const auto& distances : all_distances) {
        for (int dist : distances) {
            if (dist == INF) {
                cout << "INF ";
            } else {
                cout << dist << " ";
            }
        }
        cout << endl;
    }
    
    return 0;
}
